tests/benchmarks/test_tpch_pyspark.py

```diff
@@ -0,0 +1,248 @@
+import asyncio
+import logging
+import os
+import pathlib
+import shlex
+import subprocess
+import sys
+import time
+
+import pyspark
+import pytest
+from conda.cli.python_api import Commands, run_command
+from distributed.diagnostics.plugin import SchedulerPlugin, WorkerPlugin
+
+from . import tpch_pyspark_queries as queries
+
+# pyspark/hadoop/aws-java-* are extra sensitive to version changes it seems.
+# ref: https://hadoop.apache.org/docs/r3.3.4/hadoop-aws/dependency-analysis.html
+# for verifying aws-java-sdk-bundle version w/ hadoop version.
+assert pyspark.__version__ == "3.4.1"
+HADOOP_AWS_VERSION = "3.3.4"  # sc._jvm.org.apache.hadoop.util.VersionInfo.getVersion()
+AWS_JAVA_SDK_BUNDLE_VERSION = "1.12.262"
+
+DATASETS = {
+    "scale 100": "s3a://coiled-runtime-ci/tpch_scale_100/",
+    "scale 1000": "s3a://coiled-runtime-ci/tpch-scale-1000/",
+}
+
+
+ENABLED_DATASET = os.getenv("TPCH_SCALE")
+if ENABLED_DATASET is not None:
+    if ENABLED_DATASET not in DATASETS:
+        raise ValueError("Unknown tpch dataset: ", ENABLED_DATASET)
+else:
+    ENABLED_DATASET = "scale 1000"
+
+
+logger = logging.getLogger("coiled")
+
+
+@pytest.mark.parametrize(
+    "name",
+    sorted(
+        [name for name in dir(queries) if name.startswith("q")],
+        key=lambda v: int(v.replace("q", "")),
+    ),
+)
+def test_tpch_pyspark(tpch_pyspark_client, name):
+    from .tpch_pyspark_queries.utils import get_or_create_spark
+
+    module = getattr(queries, name)
+
+    def _retry_s3_forbidden(func):
+        # Sporadic S3 403 forbidden error
+        for i in range(3):
+            try:
+                return func()
+            except Exception as e:
+                if "403" in str(e):
+                    logger.warning("Encountered S3 403 Forbidden, retrying.")
+                    time.sleep(5)
+                    continue
+                else:
+                    raise
+        return func()
+
+    async def _run_tpch(dask_scheduler):
+        # Connecting to the Spark Connect server doens't appear to work very well
+        # If we just submit this stuff to the scheduler and generate a SparkSession there
+        # that connects to the running master we can at least get some code running
+        def _():
+            spark = get_or_create_spark(f"query{name}")
+            module.setup(spark)  # read spark tables query will select from
+
+            if hasattr(module, "ddl"):
+                spark.sql(module.ddl)  # might create temp view
+
+            q_final = spark.sql(module.query)  # benchmark query
+            try:
+                # trigger materialization of df
+                _retry_s3_forbidden(lambda: print(q_final.show(10)))
```

Comment:
by just "showing ten rows", this could trigger yet another optimization step. Naively, I would assume that this just modifies the query to have something like a `LIMIT 10` which, depending on the query, can change the result substantially.

From this perspective, to ensure the benchmarks align, we should also make sure that the dask and spark queries return the same kind of data. For instance, if the dask cluster actually submits the results back to the client, the spark cluster should do the same.
If the result is just computed and forgotten (e.g. persist + wait in dask), we should do the equivalend in spark land. Otherwise, we'll have a systematic error due to the additional client<->scheduler transmission.

Reply:
Very good point. 

In my opinion, especially if we're moving forward with benchmarking things like Polars and DuckDB, is they should at minimum compute something native to themselves, I think this is also your opinion. PySpark converting to pandas is not really encouraged, stranger still if we did that with Polars and DuckDB.

**Best approach** in my current understanding is to benchmark all in their ability to process the queries into parquet files. This is pretty unbiased, ensures each engine doesn't need to rely on some specific conversion implementation to some other engine, and means no magical lazy-non-computation occurs.

(Side note, after some experimentation, it doesn't appear `.show(...)` does anything fancy related to optimization)

Reply:
I'm not saying we should compare the results, just make sure that we're treating the final data transmission equally. I do not think we should convert anything to parquet since this just adds another IO layer that may be unreliable for the benchmark